### PR TITLE
Remove unsafe cell from SecretKey

### PR DESCRIPTION
**Pull request by davxy**
Instead of applying the resplit after the operation. Resplit **before** (by cloning where necessary).

This also allows the secret key to be `Sync`

### CODE REVIEW

File: ark-secret-scalar/src/lib.rs [hunk 1]
```diff
@@ -48,145 +42,162 @@ impl<R: RngCore+CryptoRng> XofReader for Rng2Xof<R> {
...
         selfy[1] -= &x;
     }
 
     /// Initialize and unbiased `SecretScalar` from a `XofReaader`.
     pub fn from_xof<R: XofReader>(xof: &mut R) -> Self {
         let mut xof = || xof_read_reduced(&mut *xof);
-        let mut ss = SecretScalar(UnsafeCell::new([xof(), xof()]) );
-        ss.resplit_mut();
+        let mut ss = SecretScalar([xof(), xof()]);
+        ss.resplit();
```
Comment (davxy) [inline ark-secret-scalar/src/lib.rs:103]: maybe this resplit here is redundant as the secret has just been constructed

**Comment by davxy** [comment]
Well, actually as we are constructing `SecretScalar`s on the fly (and a random split is done in the constructor) a re-split is not necessary in the SecretScalar mul and add.

Another alternative would be to define `SecretScalar` as a newtype around one single generic `PrimeField` (instead of two) and then internally having some `SecretScalarSplit` defined with two elements. And construct the `SecretScalarSplit` on the fly internally where required. Do the op and throw away the ephemeral `SecretScalarSplit` instance

**Comment by burdges** [comment]
Yeah, this should the best compromize.  I never did this before because we still interact with the same key material many times this way, but..

The big timing vulnerability is the conditional adds in the scalar multiplication, which yes this avoids just fine.  Also, we incure no costs by making the field constant time, unlike the elliptic curve operations. 

It'd be nice if arkworks provided constant-time arithmetic somehow of course, but whatever..

**Comment by davxy** [comment]
Superseeded by https://github.com/w3f/ring-vrf/pull/88

**Comment by burdges** [comment]
I kinda prefer #87 over #88 really, primarily because #87 admits more implementations.   #88 exposes the scalar directly via `pub`, not sure if that's ever useful.

The big difference is:  #88 creates one scalar directly, not two seperate ones.  If arkworks ever gets constant time arithemtic then this simplifies the spec, but SW curves are generally known for not being constant time, so if this improves the spec is debatable.

**Comment by davxy** [comment]
@burdges The public scalar thing is easily fixable: https://github.com/w3f/ring-vrf/pull/88/commits/68b45106bfce971bfda66aee5e9555d32236c9e2. Now is private. Don't know why I made it pub in the first place 😃

#88 crates one scalar but then it splits it before each operation. IMO splitting it during construction (as in #87) doesn't make sense as:
- 1. it is resplitted again before the operation (and #88 is doing it as well)
- 2. keeping it splitted in memory doesn't guarantee any additional "secrecy". As one can just add the two components.

The important thing is to use the splitted version to perform the operation, and both of them are doing it (with the difference that #88 has less footprint)

**Comment by davxy** [comment]
> We've started running schnorrkel & others in the runtime, which while dubious looks harmess. Is there a need for that here too?

If you refer to https://github.com/paritytech/polkadot-sdk/pull/2044 then this only allows to create the secret in the runtime. Signing is still not exposed by default (requires the `full_crypto` feature). This is mostly used to construct the pks for tests during genesis.

**Pull request closed without merging**